sklearn/tree/_criterion.pxd

```diff
@@ -71,6 +57,28 @@ cdef class Criterion:
     ) nogil
     cdef double proxy_impurity_improvement(self) nogil
 
+cdef class Criterion(BaseCriterion):
+    """Abstract interface for supervised impurity criteria."""
+
+    cdef const DOUBLE_t[:, ::1] y
+
+    cdef int init(
+        self,
+        const DOUBLE_t[:, ::1] y,
+        const DOUBLE_t[:] sample_weight,
+        double weighted_n_samples,
+        const SIZE_t[:] sample_indices,
+        SIZE_t start,
+        SIZE_t end
+    ) nogil except -1
+
+    cdef int pointer_init(
+        self,
+        const SIZE_t[:] sample_indices,
```

Comment:
Does `sample_indices` change every time `criterion.init` is called? If not, then there is no point?

Reply:
Thank you, Adam! I'll do the following as discussed

1. Make sure signature variables are not directly changed (e.g., y, sample_weight, weighted_n_samples, sample_indices)
2. Change downstream `pointer_reset` method